core/src/server/rpc_module.rs

```diff
@@ -500,7 +495,7 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
 	/// use jsonrpsee_core::server::RpcModule;
 	///
 	/// let mut module = RpcModule::new(());
-	/// module.register_method("say_hello", |_params, _ctx| "lo").unwrap();
+	/// module.register_method("say_hello", |_params, _connection_ctx, _ctx| "lo").unwrap();
```

Comment:
/cc @jsdw so basically this is the grumble

we expose now two context's here one `conn_ctx` which isn't exclusively connection stuff and then `ctx` which is completely controlled by the user.

however, the `ctx` is shared state per "rpc per server" and `conn_ctx` is per connection....

Reply:
Yeah ok I see!

My general grumble is that we should have some way to pass generic data from middleware or whatever into RPC calls. If we had that then we wouldn't need to expose eg the internal connection ID as a hardcoded thing, because the user could just create their own connection ID counter in middleware and pass that in.

Methods might want access to a bunch of other stuff from middleware, and I wouldn't want us to always have to work around it by trying to provide those things in a hrdcoded way :)

Reply:
For the first idea, I've been exploring with a custom type as field to the `PerConnectionContext`: `Arc<Mutex<anymap::Map<dyn anymap::any::Any + Send + Sync>>>`. This should allow users to insert in the map any type they want.

I might be missing something here, but the trick part with this comes from the fact that the `PerConnectionContext` is constructed as the last step in our `RpcServiceT` middleware. This means that users will only access it from methods, but cannot alter its state from another `RpcServiceT` middleware.
From my perspective, if the `PerConnectionContext` can only be accessed / altered from within methods, I think this brings the same usefulness as `connection: usize`.  Since `chainHead_*` methods could store internally a `Map<connection, Data>`.

https://github.com/paritytech/jsonrpsee/blob/736dd65259ba01bd31ffc8cae9d7b4d5b237d4d6/server/src/middleware/rpc/mod.rs#L42-L50

For this to be useful outside the method call; we'd probably need to extend the `call` function to include the (per connection) context : `fn call(&self, request: Request<'a>, context: PerConnectionContext) -> Self::Future;`. I do remember this was the original design for the middleware, but I think we had some difficulties implementing it?
And I think an extension of this is the second idea, however I don't know how much harder that is to implement.

Reply:
Thus, it would be neat to have a have closure similar for the context/user defined data in the rpc API such that users can decide whether the state for rpc impl should be shared or per connection...